trieste/models/keras/architectures.py

```diff
@@ -0,0 +1,315 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Callable
+from typing import Any, List, Optional, Union
+
+import numpy as np
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+from ...data import Dataset
+from .utils import get_tensor_spec_from_data
+
+
+def build_vanilla_keras_ensemble(
+    dataset: Dataset,
+    ensemble_size: int = 5,
+    num_hidden_layers: int = 2,
+    units: int = 50,
+    activation: Union[str, Callable] = "relu",
+    independent_normal: bool = False,
+) -> KerasEnsemble:
+
+    """
+    Builds a simple ensemble of neural networks in Keras where each network has the same
+    architecture: number of hidden layers, nodes in hidden layers and activation function.
+
+    :param dataset: Data for training, used for extracting input and output tensor specifications.
+    :param ensemble_size: The size of the ensemble, that is, the number of base learners or
+        individual neural networks in the ensemble.
+    :param num_hidden_layers: The number of hidden layers in each network.
+    :param units: The number of nodes in each hidden layer.
+    :param activation: The activation function in each hidden layer.
+    :param independent: If set to `True` then :class:`~tfp.layers.IndependentNormal` layer
+        is used as the output layer. This models outputs as independent, only the diagonal
+        elements of the covariance matrix are parametrized. If left as the default `False`,
+        then :class:`~tfp.layers.MultivariateNormalTriL` layer is used where correlations
+        between outputs are learned as well.
+    :return: Keras ensemble model.
+    """
+    input_tensor_spec, output_tensor_spec = get_tensor_spec_from_data(dataset)
+
+    hidden_layer_args = []
+    for i in range(num_hidden_layers):
+        hidden_layer_args.append({"units": units, "activation": activation})
+
+    networks = [
+        GaussianNetwork(
+            input_tensor_spec,
+            output_tensor_spec,
+            hidden_layer_args,
+            independent_normal,
+        )
+        for _ in range(ensemble_size)
+    ]
+    keras_ensemble = KerasEnsemble(networks)
+
+    return keras_ensemble
+
+
+class KerasEnsemble(tf.Module):
+    """
+    This class builds an ensemble of neural networks, using Keras. Individual networks must
+    be instance of :class:`~trieste.models.keras_networks.KerasEnsembleNetwork`. This class
+    is meant to be used with :class:`~trieste.models.keras_networks.DeepEnsemble` model wrapper,
+    which compiles the model.
+    """
+
+    def __init__(
+        self,
+        networks: List[KerasEnsembleNetwork],
```

Comment:
Good practice to allow any type of sequence as input (e.g. a tuple). As long as you can iterate over it and check its length.

```suggestion
        networks: Sequence[KerasEnsembleNetwork],
```